Sources/Classes/SMObjectDependencyGraph.swift

```diff
@@ -28,74 +35,118 @@ class SMObjectDependencyGraph {
       }
     }
     
-    func recursivelyMakeDependencyChains(chain: [String]) -> [[String]] {
-      if let last = chain.last, let dependencies = dependencyGraph[last], dependencies.count > 0 {
-        var chains = [[String]]()
-        for dependency in dependencies {
-          var augmentedChain = chain
-          if chain.contains(dependency) {
-            print("WARNING Loop Detected! Path \(chain) already contains '\(dependency)'")
-            chains.append(chain)
-          } else {
-            augmentedChain.append(dependency)
-            for chainFromNextRecursiveIteration in recursivelyMakeDependencyChains(chain: augmentedChain) {
-              chains.append(chainFromNextRecursiveIteration)
+    for entityName in dependencyGraph.keys {
+      ancestorGraph[entityName] = SMObjectDependencyGraph.recursivelyMakeDependencyChains(chain: [entityName], dependencyGraph: dependencyGraph)
+    }
+  }
+  
+  init(objects: [NSManagedObject]) {
+    unsorted = objects
+    //var dependencyGraph: [String:Set<String>] = [:]
+    
+    for o in objects {
+      if let entityName = o.entity.name, dependencyGraph[entityName] == nil {
+        dependencyGraph[entityName] = Set<String>()
+        for property in o.entity.properties {
+          if let relationship = property as? NSRelationshipDescription, !property.isOptional {
+            if let destinationName = relationship.destinationEntity?.name {
+              dependencyGraph[entityName]?.insert(destinationName)
             }
           }
         }
-        return chains
-      } else {
-        return [chain]
       }
     }
     
-    
     for entityName in dependencyGraph.keys {
-      ancestorGraph[entityName] = recursivelyMakeDependencyChains(chain: [entityName])
+      ancestorGraph[entityName] = SMObjectDependencyGraph.recursivelyMakeDependencyChains(chain: [entityName], dependencyGraph: dependencyGraph)
     }
   }
   
-  var sorted: [NSManagedObject] {
-    let results = unsorted.sorted { (o1:NSManagedObject, o2:NSManagedObject) -> Bool in
-      if let entityName1 = o1.entity.name, let chains1 = ancestorGraph[entityName1] {
-        if let entityName2 = o2.entity.name, let chains2 = ancestorGraph[entityName2] {
-          for chain2 in chains2 {
-            if chain2.contains(entityName1) {
-              // o2 is dependent on o1
-              //print("\(entityName2) depends on \(entityName1)")
-              return true
-            }
-          }
-          for chain1 in chains1 {
-            if chain1.contains(entityName2) {
-              // o1 is dependent on o2
-              //print("\(entityName1) depends on \(entityName1)")
-              return false
-            }
+  var sorted: [Any] {
```

Comment:
The counter argument is that by splitting the logic into other components (2 `extensions` + 1 `protocol`) you will actually make the code harder to read (though this is somewhat mitigated by the fact the new `protocol` wouldn't be exposed outside of the `SMObjectDependencyGraph` class). On the other hand your approach would indeed produce a more semantic code.

It's a tradeoff. I like my code to be linear, even at the cost of some clumsiness - but it's your repository and I will follow your lead: just let me know if you still prefer your approach after hearing out my argument!

Reply:
I just feel that `Any` is a bit of a hack.  Strong-typing is a foundation of Swift.
I don't see how it makes the code any harder to read; in fact lets you eliminate 10 or so lines of code from `var sorted:`; this is already a long 'method'.  It would also avoid the need to use implicitly unwrapped optionals for the entity names and instance identifiers.

The two different initialisers are still a concern with this approach, so there is probably further scope for refactoring.

Reply:
Ok I made the changes (see the `SMObject` protocol)